### PR TITLE
update to use dartdevc as a persistent worker

### PR DESCRIPTION
**Pull request by jakemac53**
Closes https://github.com/dart-lang/dart-pad/issues/1003

cc @devoncarew @RedBrogdon @domesticmouse not sure what combination of you would like to review this :).

This uses the bazel_worker package to run a pool of persistent dartdevc tasks and schedule jobs between them. Currently it is using the default number of workers (4), but that is easily tweakable (and I don't know what makes the most sense for dartpad - if it only handles one request at a time then just 1 would be needed although it doesn't hurt to have it higher technically - it would only spin up 1 in that case still). 

Note that I did not switch to using kernel - that would be a bit more involved and doesn't really seem necessary although we could get additional wins from it.

I did a few other cleanup things as well, feel free to push back on any of those, I will leave specific comments as to why I changed them though.

I added general dev compiler benchmarks and a spinning square benchmark for all modes and got the following results for ddc:

## Before
```
[dartdevc.hello            :  181.118ms]
[dartdevc.hellohtml        :  219.658ms]
[dartdevc.sunflower        :  241.977ms]
[dartdevc.spinning_square  :  251.619ms]
```

## After

```
[dartdevc.hello            :   19.113ms]
[dartdevc.hellohtml        :   32.594ms]
[dartdevc.sunflower        :   46.399ms]
[dartdevc.spinning_square  :   41.897ms]
```

### CODE REVIEW

File: benchmark/bench.dart [hunk 1]
```diff
@@ -8,32 +8,51 @@ library services.bench;
...
   final bool json = args.contains('--json');
 
   final BenchmarkHarness harness = BenchmarkHarness(asJson: json);
 
   final FlutterWebManager flutterWebManager = FlutterWebManager(sdkPath);
+  await flutterWebManager.warmup();
+  var compiler = Compiler(sdkPath, flutterWebManager);
+
+  Logger.root.level = Level.WARNING;
+  Logger.root.onRecord.listen((LogRecord record) {
```
Comment (jakemac53) [inline benchmark/bench.dart]: This just helps with diagnosing issues when running the benchmarks (I didn't realize I had to set up the config file for flutter_web, and just got a failure with no output before)

File: benchmark/bench.dart [hunk 2]
```diff
@@ -8,32 +8,51 @@ library services.bench;
...
 import 'package:dart_services/src/flutter_web.dart';
 
-void main(List<String> args) {
+void main(List<String> args) async {
   final bool json = args.contains('--json');
 
   final BenchmarkHarness harness = BenchmarkHarness(asJson: json);
 
   final FlutterWebManager flutterWebManager = FlutterWebManager(sdkPath);
+  await flutterWebManager.warmup();
```
Comment (jakemac53) [inline benchmark/bench.dart]: This fetches the flutter web summary files etc

File: lib/src/compiler.dart [hunk 3]
```diff
@@ -185,6 +190,11 @@ class CompilationResults {
 
   /// This is true if there were no errors.
   bool get success => problems.isEmpty;
+
+  @override
+  String toString() => success
```
Comment (jakemac53) [inline lib/src/compiler.dart]: This also helps with debugging compilation errors (previous toString was just the runtime type of the exception)

File: lib/src/flutter_web.dart [hunk 4]
```diff
@@ -52,11 +52,7 @@ $_samplePackageName:lib/
   }
 
   Future<void> warmup() async {
-    try {
-      await initFlutterWeb();
-    } catch (_) {
-      // ignore any exception here
-    }
+    await initFlutterWeb();
```
Comment (jakemac53) [inline lib/src/flutter_web.dart]: This try/catch was hiding the flutter setup issues also so I removed it - but can add back in if we need it as a safety check or something? I think it should at least log a warning or severe message though.

File: benchmark/bench.dart (hunk 2 above)
Comment (domesticmouse) [reply benchmark/bench.dart]: I feel like this should be a code comment somewhere such that this knowledge doesn't disappear on submit.

File: lib/src/flutter_web.dart (hunk 4 above)
Comment (domesticmouse) [reply lib/src/flutter_web.dart]: I really dislike catch and ignore. Happy to try living without it for the time being.

File: benchmark/bench.dart (hunk 2 above)
Comment (jakemac53) [reply benchmark/bench.dart]: done

File: lib/src/flutter_web.dart (hunk 4 above)
Comment (devoncarew) [reply lib/src/flutter_web.dart]: We shouldn't propagate any exception from here - warmup is meant to be a best effort, and not stop the initialization of the server. If this call to `initFlutterWeb` fails, a future one (reacting to a user request) could succeed.

I'd catch any exception and log it. Definitely agree that silently catching and ignoring is a good place for bugs to hide.

File: benchmark/bench.dart [hunk 5]
```diff
@@ -8,32 +8,52 @@ library services.bench;
...
 
-void main(List<String> args) {
+void main(List<String> args) async {
   final bool json = args.contains('--json');
 
   final BenchmarkHarness harness = BenchmarkHarness(asJson: json);
 
   final FlutterWebManager flutterWebManager = FlutterWebManager(sdkPath);
+  // Fetches the flutter web summary files etc from google cloud storage.
+  await flutterWebManager.warmup();
```
Comment (devoncarew) [inline benchmark/bench.dart]: If the intent here is to make the flutter web manager class ready to benchmark flutter compiles, you may want to instead call `initFlutterWeb` directly.

Comment (devoncarew) [reply benchmark/bench.dart]: Which `warmup` does today, but `initFlutterWeb` is a bit more semantic.

File: lib/src/flutter_web.dart (hunk 4 above)
Comment (jakemac53) [reply lib/src/flutter_web.dart]: done, logging a warning now

File: benchmark/bench.dart (hunk 5 above)
Comment (jakemac53) [reply benchmark/bench.dart]: Done, I also removed the comment since it is more self explanitory now.

**Comment by jakemac53** [comment]
> This looks great to me. The only thing I want to check is whether the existing mechanism that kept dart-services from trying to start two compiles at once targeting the same location will also work here. Is there anything changing that could cause state from one compileDDC request to bleed over into another or cause two processes kicked off by _ddcDriver to fight over a shared root directory for the compile?

This does work slightly differently, but it has the same behavior. The difference is that since the [processes](https://github.com/dart-lang/dart-services/pull/387/files#diff-6f8db5c7de455147a9ca9a2f78b305c2R32) are persistent they all run with the same current working directory (current just the cwd of the main process, but we could create a different temp dir if you would like and run them there for additional safety).

Each compile call still creates its [own temp directory](https://github.com/dart-lang/dart-services/pull/387/files#diff-6f8db5c7de455147a9ca9a2f78b305c2R128), where it writes the [main.dart](https://github.com/dart-lang/dart-services/pull/387/files#diff-6f8db5c7de455147a9ca9a2f78b305c2R139), and we pass the [full path](https://github.com/dart-lang/dart-services/pull/387/files#diff-6f8db5c7de455147a9ca9a2f78b305c2R146) to that to dartdevc, as well as making the [output path](https://github.com/dart-lang/dart-services/pull/387/files#diff-6f8db5c7de455147a9ca9a2f78b305c2R143) be under that directory as well.

**Comment by domesticmouse** [comment]
Thanks all!

**Comment by devoncarew** [comment]
I see this issue now when using the `compileDDC` endpoint locally:

```
VM2060:52 Uncaught TypeError: Cannot read property 'main' of undefined
    at <anonymous>:52:23
    at Object.execCb (VM1379 require.js:1696)
    at Module.check (VM1379 require.js:883)
    at Module.<anonymous> (VM1379 require.js:1139)
    at VM1379 require.js:134
    at VM1379 require.js:1189
    at each (VM1379 require.js:59)
    at Module.emit (VM1379 require.js:1188)
    at Module.check (VM1379 require.js:938)
    at Module.enable (VM1379 require.js:1176)
    at Module.init (VM1379 require.js:788)
    at callGetModule (VM1379 require.js:1203)
    at Object.completeLoad (VM1379 require.js:1590)
    at HTMLScriptElement.onScriptLoad (VM1379 require.js:1717)
```

I haven't investigated to narrow down the issue.

**Comment by devoncarew** [comment]
(I haven't bisected either, to narrow it down to this PR, but suspect it is this one)

**Comment by domesticmouse** [comment]
I'm guessing we should add tests on the `compileDDC` end point right about now?

**Comment by devoncarew** [comment]
We have some to validate that the end point produces output, but not integration tests that the output can execute and produce expected results...

**Comment by domesticmouse** [comment]
I just confirmed this is the change that caused the issues seen in https://github.com/dart-lang/dart-services/pull/387#issuecomment-483061815

I'm about to file a revert for this PR.

**Comment by jakemac53** [comment]
Hmm my guess is that the module name changed or something?

**Comment by jakemac53** [comment]
I would be happy to look at it if somebody can point me at some docs for how to set up the full stack for development :).

**Comment by jakemac53** [comment]
Actually it looks like https://github.com/dart-lang/dart-pad/blob/master/CONTRIBUTING.md has some instructions, ill see if I can work out the issue here.

**Comment by jakemac53** [comment]
hmm I can't get that fully to work (doesn't seem to actually start the server)

**Comment by devoncarew** [comment]
I'm generally following those instructions. From the dart-pad directory, I run:

```
grind serve-custom-backend
```

and from the dart-services directory I run:

```
dart bin/server_dev.dart --port 8002
```

**Comment by jakemac53** [comment]
Ah using `dart bin/server_dev.dart --port 8002` works (instead of `grind serve` which doesn't seem to work for me).

I noticed it still seems to be using dart2js though - how do i switch it to use the dev compiler?

**Comment by devoncarew** [comment]
If testing with the main dartpad UI (the desktop version), you'll need to toggle the compile time flag in lib/playground.dart - `_useDDC`.

**Comment by jakemac53** [comment]
Ok - still getting some errors fetching the `dart_sdk.js` file (loading the `dart_sdk` module) but I was able to at least see the compiled artifact and this is almost certainly the problem:

```
// Exports:
return {
  ________________________tmp__dartpadESMJSE__main: ________________________tmp__dartpadESMJSE__main
};
```

Dartpad is just expecting that to look like this:

// Exports:
return {
  main: main
};
```

I think I know how to fix that (pass --module-root).

**Pull request merged**